src/lib/keyed_container.js

```diff
@@ -0,0 +1,116 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var nestedProperty = require('./nested_property');
+
+// bitmask for deciding what's updated:
+var NONE = 0;
+var NAME = 1;
+var VALUE = 2;
+var BOTH = 3;
+
+module.exports = function keyedContainer(baseObj, path, keyName, valueName) {
```

Comment:
Yeah. It's also used in group styles, for one, though that case might be slightly more complicated. Also frames, though again, slightly more complicated. It's more a matter of realizing that unless we allow custom keys in the schema, abstracting this pattern seems like the only sane way forward.